File: smallestDifference.js
// Write a function that takes in two non-empty arrays of integers. 
// The function should find the pair of numbers (one from the first array, one from the second array) 
// whose absolute difference is closest to zero. The function should return an array containing these two numbers, 
// with the number from the first array in the first position. 
// Assume that there will only be one pair of numbers with the smallest difference.

function smallestDifference(arrayOne, arrayTwo) {
  arrayOne.sort((a, b) => a - b)
  arrayTwo.sort((a, b) => a - b)

  let smallestSoFar = Infinity, result = []
  let i = 0, j = 0

  while(i < arrayOne.length || j < arrayTwo.length) {
    let currDiff = Math.abs(arrayOne[i] - arrayTwo[j])
    if(currDiff < smallestSoFar) {
      smallestSoFar = currDiff
      result = [arrayOne[i], arrayTwo[j]]
    }

    if(arrayOne[i] < arrayTwo[j]) {
      if(i < arrayOne.length - 1)
        ++i
    } else if(j < arrayTwo.length - 1) {
      ++j
    }
  }

  return result
}

File: smallestDifference.py
import math
def smallestDifference(arrayOne, arrayTwo):
  # Write your code here.
  arrayOne = sorted(arrayOne)
  arrayTwo = sorted(arrayTwo)
  smallestSoFar = math.inf
  result = []
  i = 0
  j = 0
  while i	< len(arrayOne) - 1 and j < len(arrayTwo) - 1:
    e1 = arrayOne[i]
    e2 = arrayTwo[j]
    print([e1, e2], i , j)
    currDiff = math.fabs(e1 - e2)
    if currDiff < smallestSoFar:
      smallestSoFar = currDiff
      result = [e1, e2]
    
    if e1 < e2:
      i += 1
    elif e1 > e2:
    	j += 1
   
  
  while i < len(arrayOne)  or j < len(arrayTwo):
    currDiff = math.fabs(arrayOne[i] - arrayTwo[j])
    if currDiff < smallestSoFar:
      smallestSoFar = currDiff
      result = [arrayOne[i], arrayTwo[j]]
    if i == len(arrayOne) - 1:
      j += 1
    elif j == len(arrayTwo) -1:
      i += 1
  return result

print(smallestDifference([0, 20], [21, -2]))